File: packages/python/flatwire/failure.py
"""Partial-stream failure semantics for flatwire (Python reference).

The unspoken problem with streaming a large collection over HTTP: you have
already sent ``200 OK`` and half the array when element 40,000 throws. A bare
streamed array gives the consumer no way to tell "the producer finished cleanly"
from "the connection died" from "the producer hit an error after N items."

flatwire's **checked stream** solves this with one rule: the terminal status is
written *last*. The wire is a small envelope::

    {"items":[ e0, e1, ... ],"complete":true}
    {"items":[ e0, e1, ... ],"complete":false,"error":{"message":"...","type":"..."}}

- Items stream element-by-element, so memory stays flat.
- On success the trailer is ``"complete":true``.
- If the producer catches an exception mid-stream it emits
  ``"complete":false`` plus an ``error`` object — the already-sent items remain
  valid and the consumer is told something went wrong.
- If the stream simply ends before the trailer (connection dropped, producer
  crashed), the consumer never sees ``complete`` and detects **truncation**.

The consumer distinguishes all three outcomes::

    try:
        for row in decode_checked_array(fp):
            handle(row)
    except StreamError as e:      # producer signalled failure after N items
        ...
    except TruncatedStream:       # stream ended without a terminal status
        ...
    else:                          # clean, complete stream
        ...
"""

from __future__ import annotations

import codecs
import json
from typing import Any, BinaryIO, Iterable, Iterator


class StreamError(Exception):
    """The producer finished the stream with ``complete:false``."""

    def __init__(self, error: Any):
        self.error = error
        super().__init__(str(error))


class TruncatedStream(Exception):
    """The stream ended before a terminal status was written."""


_ENCODER = json.JSONEncoder(ensure_ascii=False, separators=(",", ":"))


def encode_checked_array(items: Iterable[Any], fp: BinaryIO) -> int:
    """Stream ``items`` inside a checked envelope, writing the terminal status
    last. If iterating ``items`` raises, a ``complete:false`` trailer carrying the
    error is written so the consumer can distinguish failure from truncation.
    Returns the number of items written. Re-raises the original exception after
    writing the error trailer.
    """
    fp.write(b'{"items":[')
    count = 0
    try:
        for item in items:
            if count:
                fp.write(b",")
            fp.write(_ENCODER.encode(item).encode("utf-8"))
            count += 1
    except Exception as exc:  # noqa: BLE001 - surface any producer error on the wire
        err = {"message": str(exc), "type": type(exc).__name__}
        fp.write(b'],"complete":false,"error":')
        fp.write(_ENCODER.encode(err).encode("utf-8"))
        fp.write(b"}")
        raise
    fp.write(b'],"complete":true}')
    return count


def decode_checked_array(fp: BinaryIO, chunk_size: int = 65536) -> Iterator[Any]:
    """Yield each item from a checked envelope, then enforce the terminal status.

    Raises :class:`StreamError` if the producer signalled ``complete:false``, and
    :class:`TruncatedStream` if the stream ended before any terminal status.
    Peak memory stays bounded by the largest element plus the small trailer.
    """
    decoder = codecs.getincrementaldecoder("utf-8")()
    buf = ""
    pos = 0
    eof = False

    def more() -> bool:
        nonlocal buf, eof
        raw = fp.read(chunk_size)
        if isinstance(raw, bytes):
            if not raw:
                tail = decoder.decode(b"", final=True)
                eof = True
                if tail:
                    buf += tail
                    return True
                return False
            text = decoder.decode(raw)
        else:
            text = raw
            if not text:
                eof = True
                return False
        buf += text
        return True

    def need(n: int) -> None:
        while len(buf) - pos < n:
            if not more():
                raise TruncatedStream("stream ended mid-envelope")

    # Consume the fixed header: {"items":[
    header = '{"items":['
    while len(buf) - pos < len(header):
        if not more():
            raise TruncatedStream("stream ended before items array")
    if buf[pos:pos + len(header)] != header:
        raise ValueError("decode_checked_array: not a flatwire checked stream")
    pos += len(header)

    # Stream array elements with a persistent-cursor depth scanner.
    elem_start = pos
    depth = 0
    in_string = False
    escape = False
    saw_any = False

    while True:
        while pos < len(buf):
            ch = buf[pos]
            if in_string:
                if escape:
                    escape = False
                elif ch == "\\":
                    escape = True
                elif ch == '"':
                    in_string = False
                pos += 1
                continue
            if ch == '"':
                in_string = True
                pos += 1
            elif ch in "{[":
                depth += 1
                pos += 1
            elif ch == "]" and depth == 0:
                # End of items array. Yield a trailing element if present.
                segment = buf[elem_start:pos].strip()
                if segment:
                    yield json.loads(segment)
                pos += 1
                # Read the (small, bounded) trailer fully and parse it.
                while not eof:
                    if not more():
                        break
                trailer = buf[pos:].strip()
                _finish(trailer)
                return
            elif ch in "}]":
                depth -= 1
                pos += 1
            elif ch == "," and depth == 0:
                segment = buf[elem_start:pos].strip()
                if segment:
                    saw_any = True
                    yield json.loads(segment)
                pos += 1
                buf = buf[pos:]
                pos = 0
                elem_start = 0
            else:
                pos += 1
        if not more():
            raise TruncatedStream("stream ended inside items array")


def _finish(trailer: str) -> None:
    # trailer looks like: ,"complete":true}   or   ,"complete":false,"error":{...}}
    if not trailer:
        raise TruncatedStream("stream ended before terminal status")
    obj = json.loads("{" + trailer.lstrip(","))
    if "complete" not in obj:
        raise TruncatedStream("stream ended before terminal status")
    if not obj["complete"]:
        raise StreamError(obj.get("error", "unknown stream error"))

File: packages/js/failure.js
'use strict';

// Partial-stream failure semantics for flatwire (Node), matching the Python
// reference. A streamed collection is wrapped in an envelope whose terminal
// status is written LAST, so a consumer distinguishes clean completion, a
// producer error after N rows, and truncation:
//
//   {"items":[ e0, e1, ... ],"complete":true}
//   {"items":[ e0, e1, ... ],"complete":false,"error":{"message":"...","name":"..."}}
//
// See docs/FAILURE.md.

const { StringDecoder } = require('node:string_decoder');

class StreamError extends Error {
  constructor(error) {
    super(typeof error === 'object' && error && error.message ? error.message : String(error));
    this.name = 'StreamError';
    this.error = error;
  }
}
class TruncatedStreamError extends Error {
  constructor(msg) { super(msg || 'stream ended before a terminal status'); this.name = 'TruncatedStreamError'; }
}

function writeBackpressured(writable, chunk) {
  return new Promise((resolve, reject) => {
    const onError = (e) => { writable.removeListener('error', onError); reject(e); };
    writable.once('error', onError);
    const ok = writable.write(chunk);
    if (ok) { writable.removeListener('error', onError); resolve(); }
    else writable.once('drain', () => { writable.removeListener('error', onError); resolve(); });
  });
}

// Stream items inside a checked envelope. If iterating `items` throws, writes a
// complete:false trailer carrying the error, then re-throws. Returns the count.
async function encodeCheckedArray(items, writable) {
  await writeBackpressured(writable, Buffer.from('{"items":[', 'utf8'));
  let count = 0;
  try {
    for await (const item of items) {
      const prefix = count ? ',' : '';
      await writeBackpressured(writable, Buffer.from(prefix + JSON.stringify(item), 'utf8'));
      count += 1;
    }
  } catch (err) {
    const errObj = { message: err && err.message ? err.message : String(err), name: err && err.name ? err.name : 'Error' };
    await writeBackpressured(writable, Buffer.from('],"complete":false,"error":' + JSON.stringify(errObj) + '}', 'utf8'));
    throw err;
  }
  await writeBackpressured(writable, Buffer.from('],"complete":true}', 'utf8'));
  return count;
}

// Yield each item, then enforce the terminal status. Throws StreamError if the
// producer signalled complete:false, TruncatedStreamError if the stream ended
// without a terminal status.
async function* decodeCheckedArray(readable) {
  const decoder = new StringDecoder('utf8');
  let buf = '';
  let ended = false;
  const it = readable[Symbol.asyncIterator]();

  async function pull() {
    const { value, done } = await it.next();
    if (done) { ended = true; buf += decoder.end(); return false; }
    buf += Buffer.isBuffer(value) ? decoder.write(value) : String(value);
    return true;
  }

  const header = '{"items":[';
  while (buf.length < header.length) {
    if (!(await pull())) throw new TruncatedStreamError('stream ended before items array');
  }
  if (buf.slice(0, header.length) !== header) throw new Error('decodeCheckedArray: not a flatwire checked stream');
  let pos = header.length;
  let elemStart = pos;
  let depth = 0, inString = false, escape = false;

  for (;;) {
    while (pos < buf.length) {
      const ch = buf[pos];
      if (inString) {
        if (escape) escape = false;
        else if (ch === '\\') escape = true;
        else if (ch === '"') inString = false;
        pos += 1; continue;
      }
      if (ch === '"') { inString = true; pos += 1; }
      else if (ch === '{' || ch === '[') { depth += 1; pos += 1; }
      else if (ch === ']' && depth === 0) {
        const seg = buf.slice(elemStart, pos).trim();
        if (seg) yield JSON.parse(seg);
        pos += 1;
        // read the (small) trailer fully
        while (!ended) await pull();
        return finish(buf.slice(pos).trim());
      }
      else if (ch === '}' || ch === ']') { depth -= 1; pos += 1; }
      else if (ch === ',' && depth === 0) {
        const seg = buf.slice(elemStart, pos).trim();
        if (seg) yield JSON.parse(seg);
        pos += 1;
        buf = buf.slice(pos); pos = 0; elemStart = 0;
      }
      else pos += 1;
    }
    if (!(await pull())) throw new TruncatedStreamError('stream ended inside items array');
  }
}

function finish(trailer) {
  if (!trailer) throw new TruncatedStreamError('stream ended before terminal status');
  const obj = JSON.parse('{' + trailer.replace(/^,/, ''));
  if (!('complete' in obj)) throw new TruncatedStreamError('stream ended before terminal status');
  if (!obj.complete) throw new StreamError(obj.error != null ? obj.error : 'unknown stream error');
}

module.exports = { encodeCheckedArray, decodeCheckedArray, StreamError, TruncatedStreamError };
